### PR TITLE
Making vimtex compatible with "Chapterfolder" 

### PR DESCRIPTION
**Pull request by effi**
Vimtex mainfile detection, TOC building and label building used to not work with the awesome CTAN package Chapterfolder[https://www.ctan.org/pkg/chapterfolder]. 
Chapterfolder provies a nice way to create nested includes for chapters, sections , subsections etc. It keeps all paths relatives so working on a bigger project gets a lot easier to handle.

The attached pull requests adds the expected compatibility. 
It adds the critical regex expressions (all labeled with cf, so re_sec becomes re_cfsec and re_input becomes re_cfinput), and extends the parsing functions with an optional argument.
Also, it will automatically change the current vim-path to the directory of the main document.

Hope you like and approve it.

### CODE REVIEW

**Comment by lervag** [comment]
Thanks! I'll look into it more closely when I get the time, cf. #207.

File: autoload/vimtex/labels.vim [hunk 1]
```diff
@@ -121,6 +126,11 @@ function! s:gather_labels(file) " {{{1
       call extend(tac, s:gather_labels(s:gather_labels_input(line, a:file)))
       continue
     endif
+    if line =~# s:re_cfinput && !s:preamble
+      " call extend(tac, s:gather_labels(s:gather_labels_cfinput(line, a:file)))
```
Comment (lervag) [inline autoload/vimtex/labels.vim]: Please remove commented code.

File: autoload/vimtex/toc.vim [hunk 2]
```diff
@@ -322,13 +333,23 @@ function! s:parse_limits(file) " {{{1
     echoerr 'File not readable: ' . a:file
     return ''
   endif
+  " echom a:file
```
Comment (lervag) [inline autoload/vimtex/toc.vim]: Please remove commented code.

File: autoload/vimtex/toc.vim [hunk 3]
```diff
@@ -322,13 +333,23 @@ function! s:parse_limits(file) " {{{1
...
 
   for line in readfile(a:file)
+    " let s:lastpath = if a:0 > 2 ? a:2 : fnamemodify(a:file,".")
```
Comment (lervag) [inline autoload/vimtex/toc.vim]: Remove.

File: autoload/vimtex/toc.vim [hunk 4]
```diff
@@ -322,13 +333,23 @@ function! s:parse_limits(file) " {{{1
...
     if line =~# s:re_input
       call s:parse_limits(s:parse_line_input(line, a:file))
+    elseif line =~# s:re_cfinput
+      call s:parse_limits(s:parse_line_input(line, a:file, "cf"))
     elseif line =~# s:re_sec
       let s:max_level = max([s:max_level,
             \ s:sec_to_value[matchstr(line, s:re_sec_level)]])
+    " elseif line =~# s:re_cfsec
+    "   let s:max_level = max([s:max_level,
+    "         \ s:sec_to_value[matchstr(line, s:re_cfsec_level)]])
```
Comment (lervag) [inline autoload/vimtex/toc.vim]: Remove.

File: autoload/vimtex/toc.vim [hunk 5]
```diff
@@ -368,6 +389,12 @@ function! s:parse_file(file) " {{{1
       call extend(toc, s:parse_file(s:parse_line_input(line, a:file)))
       continue
     endif
+    if line =~# s:re_cfinput && !s:number.preamble
+      " echom line
```
Comment (lervag) [inline autoload/vimtex/toc.vim]: Remove.

File: autoload/vimtex/toc.vim [hunk 6]
```diff
@@ -368,6 +389,12 @@ function! s:parse_file(file) " {{{1
...
+      call add(toc, s:parse_line_sec(a:file, lnum, line, "cf"))
+      call extend(toc, s:parse_file(s:parse_line_input(line, a:file, "cf")))
+      " continue
```
Comment (lervag) [inline autoload/vimtex/toc.vim]: Remove.

File: autoload/vimtex.vim [hunk 7]
```diff
@@ -156,6 +156,7 @@ function! s:init_buffer() " {{{1
 
     call add(g:vimtex_data, b:vimtex)
     let b:vimtex_id = len(g:vimtex_data) - 1
+    execute 'cd' fnameescape(b:vimtex.root)
```
Comment (lervag) [inline autoload/vimtex.vim]: No, I do not agree that we should change the current directory. This should be up to the vim user, not to the plugin. If you feel strongly about this, then I might consider to add a hook that allows a user to run commands after initialization.

File: autoload/vimtex.vim [hunk 8]
```diff
@@ -400,6 +401,12 @@ function! s:get_main_recurse(file) " {{{1
           \ . fnamemodify(a:file, ':t:r') . '(\.tex)?\s*''')) > 0
       return s:get_main_recurse(l:file)
     endif
+    if len(filter(readfile(l:file), 'v:val =~ ''\v\\cf(part|chapter|section|subsection)\**'
```
Comment (lervag) [inline autoload/vimtex.vim]: This could and should be `elseif`.

File: autoload/vimtex/labels.vim [hunk 9]
```diff
@@ -16,8 +16,13 @@ function! vimtex#labels#init_script() " {{{1
...
   let s:re_input = '\v^\s*\\%(input|include)\s*\{'
   let s:re_input_file = s:re_input . '\zs[^\}]+\ze}'
+
+  let s:re_cfinput = '\v^\s*\%*\s*\\cf%(part|chapter|%(sub)*section)\*?\s*(\[[^\]]+\])*\{'
+  let s:re_cfinput_filepath = s:re_cfinput . '[^\}]+\}\{\zs[^\}]+\ze}'
+  let s:re_cfinput_file = s:re_cfinput . '[^\}]+\}\{[^\}]+\}\{\zs[^\}]+\ze}'
+
   let s:re_label = '\v\\label\{'
-  let s:re_label_title = s:re_label . '\zs.{-}\ze\}?\s*$'
+  let s:re_label_title = s:re_label . '\zs[^\}]*\ze'
```
Comment (lervag) [inline autoload/vimtex/labels.vim]: Are you sure? Why is this better? Also, does this have anything to do with the package?

**Comment by lervag** [comment]
I've added some comments to your code. Please look through them.

Additionally, I want you to either open a new pull request that contains only a single commit, or to rebase the present pull request. I see no reason for a merge and a refactor-commit in a pull request.

Else things does look quite good. I would not mind if you could add comments that explains why the `cf`-stuff is there. Both in `labels.vim` and in `toc.vim` and in the main `vimtex.vim` there should be a comment that explains why that code is there. As you have probably noticed, I don't use very many comments, but in this case the comments are necessary, because the code treats a very special case (adds support to a latex packages that I would call relatively unknown).

File: autoload/vimtex.vim (hunk 7 above)
Comment (effi) [inline autoload/vimtex.vim]: Sorry, forgot to take this out of the commit, will rework :)

**Comment by Konfekt** [comment]
Never heard of this package either, but perhaps undeservedly though as it looks useful for bigger projects.

**Comment by lervag** [comment]
@effi Due to the long time since the last update, I've chosen to close this pull request. Please feel free to resubmit or open a new issue with a feature request.

In case you want to resubmit, please take a note of my last comment.

**Pull request closed without merging**